src/analysis/inner_product_space/finite_dimensional.lean

```diff
@@ -0,0 +1,537 @@
+/-
+Copyright (c) 2023 Monica Omar. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Monica Omar
+-/
+import analysis.inner_product_space.adjoint
+import analysis.inner_product_space.spectrum
+
+/-!
+# Finite-dimensional inner product spaces
+
+In this file, we prove some results in finite-dimensional inner product spaces.
+
+## Notation
+
+This file uses the local notation `P _` for `orthogonal_projection _`
+and `↥P _` for the extended orthogonal projection `V →L[ℂ] V`.
+
+We let `V` be an inner product space over `ℂ`.
+-/
+
+variables {V : Type*} [inner_product_space ℂ V]
+local notation `P` := orthogonal_projection
+
+/-- `U` is `T` invariant: `∀ u : V`, if `u ∈ U` then `T u ∈ U`-/
+def submodule.invariant (U : submodule ℂ V) (T : V →ₗ[ℂ] V) : Prop := U ≤ U.comap T
```

Comment:
This definition has nothing specific to inner product spaces and to complex numbers: it makes sense for any vector space over any field (or even any module over any ring), so it should be stated in this generality.

Reply:
Should I make a new file linear_algebra/invariant_submodule and put it there?

Reply:
I think this should be added in `linear_algebra/basic` as `submodule.is_invariant`, and generalized to `semilinear_map_class`. This would look something like this:
```lean
def submodule.is_invariant {R M H : Type*} [semiring R] [add_comm_monoid M] [module R M] {σ : R →+* R}
  [semilinear_map_class H σ M M] (U : submodule R M) (f : H) : Prop :=
U ≤ comap f U
```

Reply:
Oh I see you already named it `submodule.invariant`, which is fine too. But adding a new file is a bit overkill imo